core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts

```diff
@@ -30,13 +30,35 @@ export interface OpportunityDict {
   'skill_description': string;
 }
 
-interface FetchSuggestionsResponse {
+// Encapsulates the state necessary to fetch a particular suggestion from the
+// backend.
+class SuggestionFetcher {
+  // Type of suggestion to fetch.
+  type: string;
+  // The current offset, i.e. the number of items to skip in the next fetch.
+  offset: number;
+  // Cache of suggestions.
+  suggestionIdToDetails: SuggestionDetailsDict;
+
+  constructor(type: string) {
+    this.type = type;
+    this.offset = 0;
+    this.suggestionIdToDetails = {};
```

Comment:
I might be misunderstanding something here, but how do you maintain the order of already-fetched results (which needs to be stable) if you're just storing the results in an (unordered) dict?

Reply:
Can you elaborate what you mean by the order and its stability?

Reply:
If we want to show all items in the original order they are offset in the backend query result, I would have expected that we'd be maintaining a list here with the cached results in the same order we got them. This means that search results 1-10 will correspond to offsets 0-9, search results 11-20 will correspond to offsets 10-19, etc. And if you page back and forth between the pages, then the search results for 1-10, 11-20, etc. will always be the same.

But here we seem to be storing them in an unordered dict so I'm a bit confused about how we are preserving this information. Why not use a list?

Reply:
We return the results here as a dict because that's what the `contribution-and-review` component expects (it performs some lookups by suggestion). The `opportunities-list` component that actually manages the pagination does use a list in its implementation, and so order will be preserved there.